### PR TITLE
Bump autoscaler to 1.2.1

### PR DESCRIPTION
**Pull request by bikramnehra**
## Description

Ref: https://github.com/SUSE/scf/pull/2364

**JIRA:** https://jira.suse.com/browse/CAP-251

## Test plan

1. Make sure the build is passing
2. Test upgrades from CAP 1.4
3. Must be tested on CaaSP as per IBM

### CODE REVIEW

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 1]
```diff
@@ -2154,45 +2171,62 @@ instance_groups:
...
       bosh_containerization:
+        colocated_containers:
+        - loggregator-agent
         run:
           scaling:
             min: 1
             max: 3
             ha: 2
-          memory: 128
+          memory: 1024
```
Comment (bikramnehra) [inline container-host-files/etc/scf/config/role-manifest.yml]: What is the reason behind such a significant change in memory requirements?

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 2]
```diff
@@ -3128,78 +3033,13 @@ variables:
...
-    description: A PEM-encoded TLS key for clients to connect to the Autoscaler API.
-      This includes the Autoscaler ApiServer and the Service Broker.
-    required: true
 - name: AUTOSCALER_ASAPI_PUBLIC_SERVER_CERT
   options:
     secret: true
     ca: INTERNAL_CA_CERT
     alternative_names:
     - autoscaler.{{.DOMAIN}}
+    - localhost
```
Comment (bikramnehra) [inline container-host-files/etc/scf/config/role-manifest.yml]: Do we need to have `localhost` in SAN?

File: src/scf-release/src/acceptance-tests-brain/test-resources/policy.json [hunk 3]
```diff
@@ -1,32 +1,13 @@
...
+  "scaling_rules": [{
+      "metric_type": "memoryused",
+      "stat_window_secs": 60,
+      "breach_duration_secs": 60,
+      "threshold": 10,
+      "operator": ">=",
+      "cool_down_secs": 300,
+      "adjustment": "+1"
+  }]
+}
```
Comment (bikramnehra) [inline src/scf-release/src/acceptance-tests-brain/test-resources/policy.json]: newline?

File: container-host-files/etc/scf/config/opinions.yml [hunk 4]
```diff
@@ -48,100 +48,97 @@ properties:
     cf:
       client_id: app_autoscaler
       grant_type: client_credentials
-      password: "" # unused
-      username: "" # unused
+      skip_ssl_validation: true
```
Comment (jandubois) [inline container-host-files/etc/scf/config/opinions.yml]: Why is this necessary?

File: container-host-files/etc/scf/config/opinions.yml [hunk 5]
```diff
@@ -452,7 +449,7 @@ properties:
       # go into the SCF zone.
       app_autoscaler:
         # Requires read-only admin to look up stats, and write admin to do scaling
-        authorities: cloud_controller.read,cloud_controller.write,cloud_controller.admin
+        authorities: cloud_controller.read,cloud_controller.write,cloud_controller.admin,uaa.resource
```
Comment (jandubois) [inline container-host-files/etc/scf/config/opinions.yml]: I don't know what the `uaa.resource` scope does; why does the autoscaler need it?

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 6]
```diff
@@ -2154,45 +2171,62 @@ instance_groups:
...
-          scaling:
-            min: 1
-            max: 3
-            ha: 2
-          memory: 2350
-          virtual-cpus: 2
+      bosh_containerization:    
         ports:
-        - name: scaling-engine  # scalingengineport
+        - name: eventgen  # eventgenereatorport
```
Comment (jandubois) [inline container-host-files/etc/scf/config/role-manifest.yml]: Spelling: `eventgeneratorport`

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 7]
```diff
@@ -2154,45 +2171,62 @@ instance_groups:
...
+        - name: eghealthport
+          protocol: TCP
+          external: 9105
+          internal: 9105
+  - name: bpm
+    release: bpm
+    properties:
+      bosh_containerization:
+        run:
+          capabilities: [ALL]
```
Comment (jandubois) [inline container-host-files/etc/scf/config/role-manifest.yml]: No need to specify `capabilities` for a `privileged` container.

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 8]
```diff
@@ -2206,110 +2240,71 @@ instance_groups:
...
-            max: 3
-            ha: 2
-          memory: 2350
-          virtual-cpus: 2
-        ports:
-        - name: eventgen  # eventgenereatorport
-          protocol: TCP
-          external: 7105
-          internal: 7105
+          capabilities: [ALL]
```
Comment (jandubois) [inline container-host-files/etc/scf/config/role-manifest.yml]: No need to specify `capabilities` for a `privileged` container.

File: container-host-files/etc/scf/config/role-manifest.yml [hunk 9]
```diff
@@ -3258,218 +3099,29 @@ variables:
...
     required: true
-- name: AUTOSCALER_DATABASE_MAX_CONNECTIONS
+- name: AUTOSCALER_DB_ADDRESS
   options:
-    default: 1000
-    description: The maximum number of connections the Autoscaler postgres database
-      server could serve.
+    default: "autoscaler-postgres-postgres.((KUBERNETES_NAMESPACE)).svc.((KUBERNETES_CLUSTER_DOMAIN))"
+    description: The password for the Autoscaler postgres database.
+    required: true
```
Comment (jandubois) [inline container-host-files/etc/scf/config/role-manifest.yml]: The `descriptions` sounds wrong.

And `required: true` is meaningless when you have a `default` value.

File: container-host-files/etc/scf/config/role-manifest.yml (hunk 2 above)
Comment (jandubois) [inline container-host-files/etc/scf/config/role-manifest.yml]: I'm a little surprised how many certs need a `localhost` alt name. Not really a problem, but I would like to understand why this was done.

File: container-host-files/etc/scf/config/opinions.yml (hunk 4 above)
Comment (cdlliuy) [reply container-host-files/etc/scf/config/opinions.yml]: autoscaler will connect to uaa api for authorization and cf api to get app info and trigger scale.  When running with the self-signed certs ,   we need this flag to ensure the correctly connection .

File: container-host-files/etc/scf/config/opinions.yml (hunk 5 above)
Comment (cdlliuy) [reply container-host-files/etc/scf/config/opinions.yml]: this is due to a functional change in app-autoscaler in which we need to allow cf admin's access to autoscaler resourc.  To validate admin role,  the  uaa.resource authority is required to invoke /token endpoint.

File: container-host-files/etc/scf/config/opinions.yml (hunk 4 above)
Comment (jandubois) [reply container-host-files/etc/scf/config/opinions.yml]: Maybe I'm confused, but doesn't this just mean that you are missing to authorize the internal CA in some of your roles?

```
  - name: authorize-internal-ca # needs to be second as it is a dependency of other jobs.
    release: scf
```

Assuming you include this job in all your roles, all self-signed certs should verify correctly.

File: container-host-files/etc/scf/config/role-manifest.yml (hunk 1 above)
Comment (cdlliuy) [reply container-host-files/etc/scf/config/role-manifest.yml]: In the recent app-autoscaler releases,  we enabled local cache to reduce the frequency of database access for metric data, so more memory is required here.

File: container-host-files/etc/scf/config/role-manifest.yml (hunk 2 above)
Comment (cdlliuy) [reply container-host-files/etc/scf/config/role-manifest.yml]: To colocated different jobs inside 1 container and avoid the hairpin issue, 
 the "localhost" access is hardcoded, such as , https://github.com/SUSE/scf/blob/507d14f3cb354d35de47fb673e43a8e160268717/container-host-files/etc/scf/config/opinions.yml#L135
Without "localhost" in SAN, the tls connection will be rejected with dns mismatch.

**Comment by cdlliuy** [comment]
I pushed an update in https://github.com/cdlliuy/scf/commit/fed4a57329c8dee558cc186435533c177e117759 since I can't push directly in this repo. 

But given the test on my vagrant box is too slow, I didn't finish the "vagrant-preq" step yet.  That means the changes are not validated ..

File: container-host-files/etc/scf/config/role-manifest.yml (hunk 2 above)
Comment (bikramnehra) [reply container-host-files/etc/scf/config/role-manifest.yml]: already answered in previous comment so resolving this one

**Comment by bikramnehra** [comment]
@cdlliuy Unfortunately again started seeing issues in `autoscaler-brains` after pulling your latest changes. I didn't had a chance to go though the failures yet. However, here is detailed log: [autoscaler-logs.txt](https://github.com/SUSE/scf/files/3259157/autoscaler-logs.txt)

File: container-host-files/etc/scf/config/opinions.yml (hunk 4 above)
Comment (cdlliuy) [reply container-host-files/etc/scf/config/opinions.yml]: @jandubois ,  I need to keep 'skip_ssl_validation' here since the access target is the CC public API,  not for internal CA. Just like the "cf cli" and "smoke" test need to use "skip_ssl_valication" flag ..   If not,   autoscaler will report issue 
```
{"timestamp":"2019-06-06T05:01:53.211Z","source":"autoscaler:apiserver","text":"Failed to get Cloud Foundry API information","log_level":"error","data":{"info url":"https://api.cf.svc.cluster.local:9024/v2/info","error":{"code":"UNABLE_TO_VERIFY_LEAF_SIGNATURE","statusCode":500}}}
```

**Comment by cdlliuy** [comment]
@bikramnehra , I pushed a new commit https://github.com/cdlliuy/scf/commit/a83830bcbc59eca167b4f47db7743268336b9be4  to add "skip_ssl_validation" back to pass the test.   Please help to retry.
